addons/addon-base-raas/packages/base-raas-services/lib/data-source/access-strategy/roles-only/filesystem-role-service.js

```diff
@@ -0,0 +1,553 @@
+    //     in the data source account for the fs role, then update the database, see item #3 above.
+    // - If no fs role entity is found, or the existing on has its trust property full, we need to create a new
+    //   fs role as follows:
+    //   - Create the fs role entity and add the member account to the trust property, see item #3 above.
+    //     - At this point, we don't store the entity in the data base until after we are able to create
+    //       it in the data source account
+    //   - Assume the app role and create the fs role in the data source account
+    //   - Store the entity in the database
+    //   - Add usage of the study by this role, see item #1 above.
+
+    // Get app role entity
+    const appRoleEntity = await appRoleService.mustFind(requestContext, { accountId, bucket, arn: appRoleArn });
+
+    // Get all existing usages by fs roles for this study (keeping env permission in mind)
+    const fsRoleUsages = await resourceUsageService.getResourceUsage(requestContext, {
+      resource: studyResourceId,
+      setName: fsUsageSetName,
+    });
+    const fsRoleEntities = _.get(fsRoleUsages, fsUsageSetName, []);
+
+    // Start a loop to see if existing fs roles for the study have the member account already
+    // or can take it without exceeding the trust doc limit?
+    let fsRoleEntity;
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const fsRoleArn of fsRoleEntities) {
+      const entity = await this.find(requestContext, { accountId, bucket, arn: fsRoleArn });
+      // eslint-disable-next-line no-continue
+      if (!entity) continue;
+
+      // Check if the provided study matches any of the studies in the fs role entity, including the env permissions
+      // eslint-disable-next-line no-continue
+      if (!hasStudy(entity, studyEntity)) continue;
+
+      if (hasMemberAccount(entity, memberAccountId)) {
+        // We found an existing fs role entity that already has the member account
+        fsRoleEntity = entity;
+
+        // Add environment usage
+        await resourceUsageService.addUsage(requestContext, {
+          resource: studyResourceId,
+          setName: memberAccountId,
+          item: envId,
+        });
+
+        break;
+      }
+
+      if (!maxReached(addMemberAccount(entity, memberAccountId))) {
+        // We found an existing fs role entity and we can fit the member account in its trust doc
+
+        // Update the fs role policy in the data source account
+        await this.updateAssumeRolePolicy(entity);
+
+        // Add environment usage
+        await resourceUsageService.addUsage(requestContext, {
+          resource: studyResourceId,
+          setName: memberAccountId,
+          item: envId,
+        });
+
+        // We need to save the fsRoleEntity (even though it is already in the database), because
+        // we added the member account id to the trust doc.
+        fsRoleEntity = await this.saveEntity(requestContext, entity);
```

Comment:
If this save fails, can the method be retired successfully to save again?

Reply:
I agonized over this when I was writing the code :)
Introducing a transactional behavior to this method means a much much more complex approach, because we are doing mutating calls using AWS IAM roles, it means that we need to consider the command pattern with undo logic.  For now, I went with this approach while acknowledging its shortcomings.